<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta name="generator" content="Asciidoctor 2.0.12">
<title>Flu-crew Technical Documentation</title>
<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:300,300italic,400,400italic,600,600italic%7CNoto+Serif:400,400italic,700,700italic%7CDroid+Sans+Mono:400,700">
<style>
/* Asciidoctor default stylesheet | MIT License | https://asciidoctor.org */
/* Uncomment @import statement to use as custom stylesheet */
/*@import "https://fonts.googleapis.com/css?family=Open+Sans:300,300italic,400,400italic,600,600italic%7CNoto+Serif:400,400italic,700,700italic%7CDroid+Sans+Mono:400,700";*/
article,aside,details,figcaption,figure,footer,header,hgroup,main,nav,section{display:block}
audio,video{display:inline-block}
audio:not([controls]){display:none;height:0}
html{font-family:sans-serif;-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%}
a{background:none}
a:focus{outline:thin dotted}
a:active,a:hover{outline:0}
h1{font-size:2em;margin:.67em 0}
abbr[title]{border-bottom:1px dotted}
b,strong{font-weight:bold}
dfn{font-style:italic}
hr{-moz-box-sizing:content-box;box-sizing:content-box;height:0}
mark{background:#ff0;color:#000}
code,kbd,pre,samp{font-family:monospace;font-size:1em}
pre{white-space:pre-wrap}
q{quotes:"\201C" "\201D" "\2018" "\2019"}
small{font-size:80%}
sub,sup{font-size:75%;line-height:0;position:relative;vertical-align:baseline}
sup{top:-.5em}
sub{bottom:-.25em}
img{border:0}
svg:not(:root){overflow:hidden}
figure{margin:0}
fieldset{border:1px solid silver;margin:0 2px;padding:.35em .625em .75em}
legend{border:0;padding:0}
button,input,select,textarea{font-family:inherit;font-size:100%;margin:0}
button,input{line-height:normal}
button,select{text-transform:none}
button,html input[type="button"],input[type="reset"],input[type="submit"]{-webkit-appearance:button;cursor:pointer}
button[disabled],html input[disabled]{cursor:default}
input[type="checkbox"],input[type="radio"]{box-sizing:border-box;padding:0}
button::-moz-focus-inner,input::-moz-focus-inner{border:0;padding:0}
textarea{overflow:auto;vertical-align:top}
table{border-collapse:collapse;border-spacing:0}
*,*::before,*::after{-moz-box-sizing:border-box;-webkit-box-sizing:border-box;box-sizing:border-box}
html,body{font-size:100%}
body{background:#fff;color:rgba(0,0,0,.8);padding:0;margin:0;font-family:"Noto Serif","DejaVu Serif",serif;font-weight:400;font-style:normal;line-height:1;position:relative;cursor:auto;tab-size:4;word-wrap:anywhere;-moz-osx-font-smoothing:grayscale;-webkit-font-smoothing:antialiased}
a:hover{cursor:pointer}
img,object,embed{max-width:100%;height:auto}
object,embed{height:100%}
img{-ms-interpolation-mode:bicubic}
.left{float:left!important}
.right{float:right!important}
.text-left{text-align:left!important}
.text-right{text-align:right!important}
.text-center{text-align:center!important}
.text-justify{text-align:justify!important}
.hide{display:none}
img,object,svg{display:inline-block;vertical-align:middle}
textarea{height:auto;min-height:50px}
select{width:100%}
.subheader,.admonitionblock td.content>.title,.audioblock>.title,.exampleblock>.title,.imageblock>.title,.listingblock>.title,.literalblock>.title,.stemblock>.title,.openblock>.title,.paragraph>.title,.quoteblock>.title,table.tableblock>.title,.verseblock>.title,.videoblock>.title,.dlist>.title,.olist>.title,.ulist>.title,.qlist>.title,.hdlist>.title{line-height:1.45;color:#7a2518;font-weight:400;margin-top:0;margin-bottom:.25em}
div,dl,dt,dd,ul,ol,li,h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6,pre,form,p,blockquote,th,td{margin:0;padding:0}
a{color:#2156a5;text-decoration:underline;line-height:inherit}
a:hover,a:focus{color:#1d4b8f}
a img{border:0}
p{font-family:inherit;font-weight:400;font-size:1em;line-height:1.6;margin-bottom:1.25em;text-rendering:optimizeLegibility}
p aside{font-size:.875em;line-height:1.35;font-style:italic}
h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6{font-family:"Open Sans","DejaVu Sans",sans-serif;font-weight:300;font-style:normal;color:#ba3925;text-rendering:optimizeLegibility;margin-top:1em;margin-bottom:.5em;line-height:1.0125em}
h1 small,h2 small,h3 small,#toctitle small,.sidebarblock>.content>.title small,h4 small,h5 small,h6 small{font-size:60%;color:#e99b8f;line-height:0}
h1{font-size:2.125em}
h2{font-size:1.6875em}
h3,#toctitle,.sidebarblock>.content>.title{font-size:1.375em}
h4,h5{font-size:1.125em}
h6{font-size:1em}
hr{border:solid #dddddf;border-width:1px 0 0;clear:both;margin:1.25em 0 1.1875em;height:0}
em,i{font-style:italic;line-height:inherit}
strong,b{font-weight:bold;line-height:inherit}
small{font-size:60%;line-height:inherit}
code{font-family:"Droid Sans Mono","DejaVu Sans Mono",monospace;font-weight:400;color:rgba(0,0,0,.9)}
ul,ol,dl{font-size:1em;line-height:1.6;margin-bottom:1.25em;list-style-position:outside;font-family:inherit}
ul,ol{margin-left:1.5em}
ul li ul,ul li ol{margin-left:1.25em;margin-bottom:0;font-size:1em}
ul.square li ul,ul.circle li ul,ul.disc li ul{list-style:inherit}
ul.square{list-style-type:square}
ul.circle{list-style-type:circle}
ul.disc{list-style-type:disc}
ol li ul,ol li ol{margin-left:1.25em;margin-bottom:0}
dl dt{margin-bottom:.3125em;font-weight:bold}
dl dd{margin-bottom:1.25em}
abbr,acronym{text-transform:uppercase;font-size:90%;color:rgba(0,0,0,.8);border-bottom:1px dotted #ddd;cursor:help}
abbr{text-transform:none}
blockquote{margin:0 0 1.25em;padding:.5625em 1.25em 0 1.1875em;border-left:1px solid #ddd}
blockquote cite{display:block;font-size:.9375em;color:rgba(0,0,0,.6)}
blockquote cite::before{content:"\2014 \0020"}
blockquote cite a,blockquote cite a:visited{color:rgba(0,0,0,.6)}
blockquote,blockquote p{line-height:1.6;color:rgba(0,0,0,.85)}
@media screen and (min-width:768px){h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6{line-height:1.2}
h1{font-size:2.75em}
h2{font-size:2.3125em}
h3,#toctitle,.sidebarblock>.content>.title{font-size:1.6875em}
h4{font-size:1.4375em}}
table{background:#fff;margin-bottom:1.25em;border:solid 1px #dedede;word-wrap:normal}
table thead,table tfoot{background:#f7f8f7}
table thead tr th,table thead tr td,table tfoot tr th,table tfoot tr td{padding:.5em .625em .625em;font-size:inherit;color:rgba(0,0,0,.8);text-align:left}
table tr th,table tr td{padding:.5625em .625em;font-size:inherit;color:rgba(0,0,0,.8)}
table tr.even,table tr.alt{background:#f8f8f7}
table thead tr th,table tfoot tr th,table tbody tr td,table tr td,table tfoot tr td{line-height:1.6}
h1,h2,h3,#toctitle,.sidebarblock>.content>.title,h4,h5,h6{line-height:1.2;word-spacing:-.05em}
h1 strong,h2 strong,h3 strong,#toctitle strong,.sidebarblock>.content>.title strong,h4 strong,h5 strong,h6 strong{font-weight:400}
.center{margin-left:auto;margin-right:auto}
.stretch{width:100%}
.clearfix::before,.clearfix::after,.float-group::before,.float-group::after{content:" ";display:table}
.clearfix::after,.float-group::after{clear:both}
:not(pre).nobreak{word-wrap:normal}
:not(pre).nowrap{white-space:nowrap}
:not(pre).pre-wrap{white-space:pre-wrap}
:not(pre):not([class^=L])>code{font-size:.9375em;font-style:normal!important;letter-spacing:0;padding:.1em .5ex;word-spacing:-.15em;background:#f7f7f8;-webkit-border-radius:4px;border-radius:4px;line-height:1.45;text-rendering:optimizeSpeed}
pre{color:rgba(0,0,0,.9);font-family:"Droid Sans Mono","DejaVu Sans Mono",monospace;line-height:1.45;text-rendering:optimizeSpeed}
pre code,pre pre{color:inherit;font-size:inherit;line-height:inherit}
pre>code{display:block}
pre.nowrap,pre.nowrap pre{white-space:pre;word-wrap:normal}
em em{font-style:normal}
strong strong{font-weight:400}
.keyseq{color:rgba(51,51,51,.8)}
kbd{font-family:"Droid Sans Mono","DejaVu Sans Mono",monospace;display:inline-block;color:rgba(0,0,0,.8);font-size:.65em;line-height:1.45;background:#f7f7f7;border:1px solid #ccc;-webkit-border-radius:3px;border-radius:3px;-webkit-box-shadow:0 1px 0 rgba(0,0,0,.2),0 0 0 .1em white inset;box-shadow:0 1px 0 rgba(0,0,0,.2),0 0 0 .1em #fff inset;margin:0 .15em;padding:.2em .5em;vertical-align:middle;position:relative;top:-.1em;white-space:nowrap}
.keyseq kbd:first-child{margin-left:0}
.keyseq kbd:last-child{margin-right:0}
.menuseq,.menuref{color:#000}
.menuseq b:not(.caret),.menuref{font-weight:inherit}
.menuseq{word-spacing:-.02em}
.menuseq b.caret{font-size:1.25em;line-height:.8}
.menuseq i.caret{font-weight:bold;text-align:center;width:.45em}
b.button::before,b.button::after{position:relative;top:-1px;font-weight:400}
b.button::before{content:"[";padding:0 3px 0 2px}
b.button::after{content:"]";padding:0 2px 0 3px}
p a>code:hover{color:rgba(0,0,0,.9)}
#header,#content,#footnotes,#footer{width:100%;margin-left:auto;margin-right:auto;margin-top:0;margin-bottom:0;max-width:62.5em;*zoom:1;position:relative;padding-left:.9375em;padding-right:.9375em}
#header::before,#header::after,#content::before,#content::after,#footnotes::before,#footnotes::after,#footer::before,#footer::after{content:" ";display:table}
#header::after,#content::after,#footnotes::after,#footer::after{clear:both}
#content{margin-top:1.25em}
#content::before{content:none}
#header>h1:first-child{color:rgba(0,0,0,.85);margin-top:2.25rem;margin-bottom:0}
#header>h1:first-child+#toc{margin-top:8px;border-top:1px solid #dddddf}
#header>h1:only-child,body.toc2 #header>h1:nth-last-child(2){border-bottom:1px solid #dddddf;padding-bottom:8px}
#header .details{border-bottom:1px solid #dddddf;line-height:1.45;padding-top:.25em;padding-bottom:.25em;padding-left:.25em;color:rgba(0,0,0,.6);display:-ms-flexbox;display:-webkit-flex;display:flex;-ms-flex-flow:row wrap;-webkit-flex-flow:row wrap;flex-flow:row wrap}
#header .details span:first-child{margin-left:-.125em}
#header .details span.email a{color:rgba(0,0,0,.85)}
#header .details br{display:none}
#header .details br+span::before{content:"\00a0\2013\00a0"}
#header .details br+span.author::before{content:"\00a0\22c5\00a0";color:rgba(0,0,0,.85)}
#header .details br+span#revremark::before{content:"\00a0|\00a0"}
#header #revnumber{text-transform:capitalize}
#header #revnumber::after{content:"\00a0"}
#content>h1:first-child:not([class]){color:rgba(0,0,0,.85);border-bottom:1px solid #dddddf;padding-bottom:8px;margin-top:0;padding-top:1rem;margin-bottom:1.25rem}
#toc{border-bottom:1px solid #e7e7e9;padding-bottom:.5em}
#toc>ul{margin-left:.125em}
#toc ul.sectlevel0>li>a{font-style:italic}
#toc ul.sectlevel0 ul.sectlevel1{margin:.5em 0}
#toc ul{font-family:"Open Sans","DejaVu Sans",sans-serif;list-style-type:none}
#toc li{line-height:1.3334;margin-top:.3334em}
#toc a{text-decoration:none}
#toc a:active{text-decoration:underline}
#toctitle{color:#7a2518;font-size:1.2em}
@media screen and (min-width:768px){#toctitle{font-size:1.375em}
body.toc2{padding-left:15em;padding-right:0}
#toc.toc2{margin-top:0!important;background:#f8f8f7;position:fixed;width:15em;left:0;top:0;border-right:1px solid #e7e7e9;border-top-width:0!important;border-bottom-width:0!important;z-index:1000;padding:1.25em 1em;height:100%;overflow:auto}
#toc.toc2 #toctitle{margin-top:0;margin-bottom:.8rem;font-size:1.2em}
#toc.toc2>ul{font-size:.9em;margin-bottom:0}
#toc.toc2 ul ul{margin-left:0;padding-left:1em}
#toc.toc2 ul.sectlevel0 ul.sectlevel1{padding-left:0;margin-top:.5em;margin-bottom:.5em}
body.toc2.toc-right{padding-left:0;padding-right:15em}
body.toc2.toc-right #toc.toc2{border-right-width:0;border-left:1px solid #e7e7e9;left:auto;right:0}}
@media screen and (min-width:1280px){body.toc2{padding-left:20em;padding-right:0}
#toc.toc2{width:20em}
#toc.toc2 #toctitle{font-size:1.375em}
#toc.toc2>ul{font-size:.95em}
#toc.toc2 ul ul{padding-left:1.25em}
body.toc2.toc-right{padding-left:0;padding-right:20em}}
#content #toc{border-style:solid;border-width:1px;border-color:#e0e0dc;margin-bottom:1.25em;padding:1.25em;background:#f8f8f7;-webkit-border-radius:4px;border-radius:4px}
#content #toc>:first-child{margin-top:0}
#content #toc>:last-child{margin-bottom:0}
#footer{max-width:none;background:rgba(0,0,0,.8);padding:1.25em}
#footer-text{color:rgba(255,255,255,.8);line-height:1.44}
#content{margin-bottom:.625em}
.sect1{padding-bottom:.625em}
@media screen and (min-width:768px){#content{margin-bottom:1.25em}
.sect1{padding-bottom:1.25em}}
.sect1:last-child{padding-bottom:0}
.sect1+.sect1{border-top:1px solid #e7e7e9}
#content h1>a.anchor,h2>a.anchor,h3>a.anchor,#toctitle>a.anchor,.sidebarblock>.content>.title>a.anchor,h4>a.anchor,h5>a.anchor,h6>a.anchor{position:absolute;z-index:1001;width:1.5ex;margin-left:-1.5ex;display:block;text-decoration:none!important;visibility:hidden;text-align:center;font-weight:400}
#content h1>a.anchor::before,h2>a.anchor::before,h3>a.anchor::before,#toctitle>a.anchor::before,.sidebarblock>.content>.title>a.anchor::before,h4>a.anchor::before,h5>a.anchor::before,h6>a.anchor::before{content:"\00A7";font-size:.85em;display:block;padding-top:.1em}
#content h1:hover>a.anchor,#content h1>a.anchor:hover,h2:hover>a.anchor,h2>a.anchor:hover,h3:hover>a.anchor,#toctitle:hover>a.anchor,.sidebarblock>.content>.title:hover>a.anchor,h3>a.anchor:hover,#toctitle>a.anchor:hover,.sidebarblock>.content>.title>a.anchor:hover,h4:hover>a.anchor,h4>a.anchor:hover,h5:hover>a.anchor,h5>a.anchor:hover,h6:hover>a.anchor,h6>a.anchor:hover{visibility:visible}
#content h1>a.link,h2>a.link,h3>a.link,#toctitle>a.link,.sidebarblock>.content>.title>a.link,h4>a.link,h5>a.link,h6>a.link{color:#ba3925;text-decoration:none}
#content h1>a.link:hover,h2>a.link:hover,h3>a.link:hover,#toctitle>a.link:hover,.sidebarblock>.content>.title>a.link:hover,h4>a.link:hover,h5>a.link:hover,h6>a.link:hover{color:#a53221}
details,.audioblock,.imageblock,.literalblock,.listingblock,.stemblock,.videoblock{margin-bottom:1.25em}
details>summary:first-of-type{cursor:pointer;display:list-item;outline:none;margin-bottom:.75em}
.admonitionblock td.content>.title,.audioblock>.title,.exampleblock>.title,.imageblock>.title,.listingblock>.title,.literalblock>.title,.stemblock>.title,.openblock>.title,.paragraph>.title,.quoteblock>.title,table.tableblock>.title,.verseblock>.title,.videoblock>.title,.dlist>.title,.olist>.title,.ulist>.title,.qlist>.title,.hdlist>.title{text-rendering:optimizeLegibility;text-align:left;font-family:"Noto Serif","DejaVu Serif",serif;font-size:1rem;font-style:italic}
table.tableblock.fit-content>caption.title{white-space:nowrap;width:0}
.paragraph.lead>p,#preamble>.sectionbody>[class="paragraph"]:first-of-type p{font-size:1.21875em;line-height:1.6;color:rgba(0,0,0,.85)}
table.tableblock #preamble>.sectionbody>[class="paragraph"]:first-of-type p{font-size:inherit}
.admonitionblock>table{border-collapse:separate;border:0;background:none;width:100%}
.admonitionblock>table td.icon{text-align:center;width:80px}
.admonitionblock>table td.icon img{max-width:none}
.admonitionblock>table td.icon .title{font-weight:bold;font-family:"Open Sans","DejaVu Sans",sans-serif;text-transform:uppercase}
.admonitionblock>table td.content{padding-left:1.125em;padding-right:1.25em;border-left:1px solid #dddddf;color:rgba(0,0,0,.6);word-wrap:anywhere}
.admonitionblock>table td.content>:last-child>:last-child{margin-bottom:0}
.exampleblock>.content{border-style:solid;border-width:1px;border-color:#e6e6e6;margin-bottom:1.25em;padding:1.25em;background:#fff;-webkit-border-radius:4px;border-radius:4px}
.exampleblock>.content>:first-child{margin-top:0}
.exampleblock>.content>:last-child{margin-bottom:0}
.sidebarblock{border-style:solid;border-width:1px;border-color:#dbdbd6;margin-bottom:1.25em;padding:1.25em;background:#f3f3f2;-webkit-border-radius:4px;border-radius:4px}
.sidebarblock>:first-child{margin-top:0}
.sidebarblock>:last-child{margin-bottom:0}
.sidebarblock>.content>.title{color:#7a2518;margin-top:0;text-align:center}
.exampleblock>.content>:last-child>:last-child,.exampleblock>.content .olist>ol>li:last-child>:last-child,.exampleblock>.content .ulist>ul>li:last-child>:last-child,.exampleblock>.content .qlist>ol>li:last-child>:last-child,.sidebarblock>.content>:last-child>:last-child,.sidebarblock>.content .olist>ol>li:last-child>:last-child,.sidebarblock>.content .ulist>ul>li:last-child>:last-child,.sidebarblock>.content .qlist>ol>li:last-child>:last-child{margin-bottom:0}
.literalblock pre,.listingblock>.content>pre{-webkit-border-radius:4px;border-radius:4px;overflow-x:auto;padding:1em;font-size:.8125em}
@media screen and (min-width:768px){.literalblock pre,.listingblock>.content>pre{font-size:.90625em}}
@media screen and (min-width:1280px){.literalblock pre,.listingblock>.content>pre{font-size:1em}}
.literalblock pre,.listingblock>.content>pre:not(.highlight),.listingblock>.content>pre[class="highlight"],.listingblock>.content>pre[class^="highlight "]{background:#f7f7f8}
.literalblock.output pre{color:#f7f7f8;background:rgba(0,0,0,.9)}
.listingblock>.content{position:relative}
.listingblock code[data-lang]::before{display:none;content:attr(data-lang);position:absolute;font-size:.75em;top:.425rem;right:.5rem;line-height:1;text-transform:uppercase;color:inherit;opacity:.5}
.listingblock:hover code[data-lang]::before{display:block}
.listingblock.terminal pre .command::before{content:attr(data-prompt);padding-right:.5em;color:inherit;opacity:.5}
.listingblock.terminal pre .command:not([data-prompt])::before{content:"$"}
.listingblock pre.highlightjs{padding:0}
.listingblock pre.highlightjs>code{padding:1em;-webkit-border-radius:4px;border-radius:4px}
.listingblock pre.prettyprint{border-width:0}
.prettyprint{background:#f7f7f8}
pre.prettyprint .linenums{line-height:1.45;margin-left:2em}
pre.prettyprint li{background:none;list-style-type:inherit;padding-left:0}
pre.prettyprint li code[data-lang]::before{opacity:1}
pre.prettyprint li:not(:first-child) code[data-lang]::before{display:none}
table.linenotable{border-collapse:separate;border:0;margin-bottom:0;background:none}
table.linenotable td[class]{color:inherit;vertical-align:top;padding:0;line-height:inherit;white-space:normal}
table.linenotable td.code{padding-left:.75em}
table.linenotable td.linenos{border-right:1px solid currentColor;opacity:.35;padding-right:.5em}
pre.pygments .lineno{border-right:1px solid currentColor;opacity:.35;display:inline-block;margin-right:.75em}
pre.pygments .lineno::before{content:"";margin-right:-.125em}
.quoteblock{margin:0 1em 1.25em 1.5em;display:table}
.quoteblock:not(.excerpt)>.title{margin-left:-1.5em;margin-bottom:.75em}
.quoteblock blockquote,.quoteblock p{color:rgba(0,0,0,.85);font-size:1.15rem;line-height:1.75;word-spacing:.1em;letter-spacing:0;font-style:italic;text-align:justify}
.quoteblock blockquote{margin:0;padding:0;border:0}
.quoteblock blockquote::before{content:"\201c";float:left;font-size:2.75em;font-weight:bold;line-height:.6em;margin-left:-.6em;color:#7a2518;text-shadow:0 1px 2px rgba(0,0,0,.1)}
.quoteblock blockquote>.paragraph:last-child p{margin-bottom:0}
.quoteblock .attribution{margin-top:.75em;margin-right:.5ex;text-align:right}
.verseblock{margin:0 1em 1.25em}
.verseblock pre{font-family:"Open Sans","DejaVu Sans",sans;font-size:1.15rem;color:rgba(0,0,0,.85);font-weight:300;text-rendering:optimizeLegibility}
.verseblock pre strong{font-weight:400}
.verseblock .attribution{margin-top:1.25rem;margin-left:.5ex}
.quoteblock .attribution,.verseblock .attribution{font-size:.9375em;line-height:1.45;font-style:italic}
.quoteblock .attribution br,.verseblock .attribution br{display:none}
.quoteblock .attribution cite,.verseblock .attribution cite{display:block;letter-spacing:-.025em;color:rgba(0,0,0,.6)}
.quoteblock.abstract blockquote::before,.quoteblock.excerpt blockquote::before,.quoteblock .quoteblock blockquote::before{display:none}
.quoteblock.abstract blockquote,.quoteblock.abstract p,.quoteblock.excerpt blockquote,.quoteblock.excerpt p,.quoteblock .quoteblock blockquote,.quoteblock .quoteblock p{line-height:1.6;word-spacing:0}
.quoteblock.abstract{margin:0 1em 1.25em;display:block}
.quoteblock.abstract>.title{margin:0 0 .375em;font-size:1.15em;text-align:center}
.quoteblock.excerpt>blockquote,.quoteblock .quoteblock{padding:0 0 .25em 1em;border-left:.25em solid #dddddf}
.quoteblock.excerpt,.quoteblock .quoteblock{margin-left:0}
.quoteblock.excerpt blockquote,.quoteblock.excerpt p,.quoteblock .quoteblock blockquote,.quoteblock .quoteblock p{color:inherit;font-size:1.0625rem}
.quoteblock.excerpt .attribution,.quoteblock .quoteblock .attribution{color:inherit;text-align:left;margin-right:0}
p.tableblock:last-child{margin-bottom:0}
td.tableblock>.content{margin-bottom:1.25em;word-wrap:anywhere}
td.tableblock>.content>:last-child{margin-bottom:-1.25em}
table.tableblock,th.tableblock,td.tableblock{border:0 solid #dedede}
table.grid-all>*>tr>*{border-width:1px}
table.grid-cols>*>tr>*{border-width:0 1px}
table.grid-rows>*>tr>*{border-width:1px 0}
table.frame-all{border-width:1px}
table.frame-ends{border-width:1px 0}
table.frame-sides{border-width:0 1px}
table.frame-none>colgroup+*>:first-child>*,table.frame-sides>colgroup+*>:first-child>*{border-top-width:0}
table.frame-none>:last-child>:last-child>*,table.frame-sides>:last-child>:last-child>*{border-bottom-width:0}
table.frame-none>*>tr>:first-child,table.frame-ends>*>tr>:first-child{border-left-width:0}
table.frame-none>*>tr>:last-child,table.frame-ends>*>tr>:last-child{border-right-width:0}
table.stripes-all tr,table.stripes-odd tr:nth-of-type(odd),table.stripes-even tr:nth-of-type(even),table.stripes-hover tr:hover{background:#f8f8f7}
th.halign-left,td.halign-left{text-align:left}
th.halign-right,td.halign-right{text-align:right}
th.halign-center,td.halign-center{text-align:center}
th.valign-top,td.valign-top{vertical-align:top}
th.valign-bottom,td.valign-bottom{vertical-align:bottom}
th.valign-middle,td.valign-middle{vertical-align:middle}
table thead th,table tfoot th{font-weight:bold}
tbody tr th{background:#f7f8f7}
tbody tr th,tbody tr th p,tfoot tr th,tfoot tr th p{color:rgba(0,0,0,.8);font-weight:bold}
p.tableblock>code:only-child{background:none;padding:0}
p.tableblock{font-size:1em}
ol{margin-left:1.75em}
ul li ol{margin-left:1.5em}
dl dd{margin-left:1.125em}
dl dd:last-child,dl dd:last-child>:last-child{margin-bottom:0}
ol>li p,ul>li p,ul dd,ol dd,.olist .olist,.ulist .ulist,.ulist .olist,.olist .ulist{margin-bottom:.625em}
ul.checklist,ul.none,ol.none,ul.no-bullet,ol.no-bullet,ol.unnumbered,ul.unstyled,ol.unstyled{list-style-type:none}
ul.no-bullet,ol.no-bullet,ol.unnumbered{margin-left:.625em}
ul.unstyled,ol.unstyled{margin-left:0}
ul.checklist{margin-left:.625em}
ul.checklist li>p:first-child>.fa-square-o:first-child,ul.checklist li>p:first-child>.fa-check-square-o:first-child{width:1.25em;font-size:.8em;position:relative;bottom:.125em}
ul.checklist li>p:first-child>input[type="checkbox"]:first-child{margin-right:.25em}
ul.inline{display:-ms-flexbox;display:-webkit-box;display:flex;-ms-flex-flow:row wrap;-webkit-flex-flow:row wrap;flex-flow:row wrap;list-style:none;margin:0 0 .625em -1.25em}
ul.inline>li{margin-left:1.25em}
.unstyled dl dt{font-weight:400;font-style:normal}
ol.arabic{list-style-type:decimal}
ol.decimal{list-style-type:decimal-leading-zero}
ol.loweralpha{list-style-type:lower-alpha}
ol.upperalpha{list-style-type:upper-alpha}
ol.lowerroman{list-style-type:lower-roman}
ol.upperroman{list-style-type:upper-roman}
ol.lowergreek{list-style-type:lower-greek}
.hdlist>table,.colist>table{border:0;background:none}
.hdlist>table>tbody>tr,.colist>table>tbody>tr{background:none}
td.hdlist1,td.hdlist2{vertical-align:top;padding:0 .625em}
td.hdlist1{font-weight:bold;padding-bottom:1.25em}
td.hdlist2{word-wrap:anywhere}
.literalblock+.colist,.listingblock+.colist{margin-top:-.5em}
.colist td:not([class]):first-child{padding:.4em .75em 0;line-height:1;vertical-align:top}
.colist td:not([class]):first-child img{max-width:none}
.colist td:not([class]):last-child{padding:.25em 0}
.thumb,.th{line-height:0;display:inline-block;border:solid 4px #fff;-webkit-box-shadow:0 0 0 1px #ddd;box-shadow:0 0 0 1px #ddd}
.imageblock.left{margin:.25em .625em 1.25em 0}
.imageblock.right{margin:.25em 0 1.25em .625em}
.imageblock>.title{margin-bottom:0}
.imageblock.thumb,.imageblock.th{border-width:6px}
.imageblock.thumb>.title,.imageblock.th>.title{padding:0 .125em}
.image.left,.image.right{margin-top:.25em;margin-bottom:.25em;display:inline-block;line-height:0}
.image.left{margin-right:.625em}
.image.right{margin-left:.625em}
a.image{text-decoration:none;display:inline-block}
a.image object{pointer-events:none}
sup.footnote,sup.footnoteref{font-size:.875em;position:static;vertical-align:super}
sup.footnote a,sup.footnoteref a{text-decoration:none}
sup.footnote a:active,sup.footnoteref a:active{text-decoration:underline}
#footnotes{padding-top:.75em;padding-bottom:.75em;margin-bottom:.625em}
#footnotes hr{width:20%;min-width:6.25em;margin:-.25em 0 .75em;border-width:1px 0 0}
#footnotes .footnote{padding:0 .375em 0 .225em;line-height:1.3334;font-size:.875em;margin-left:1.2em;margin-bottom:.2em}
#footnotes .footnote a:first-of-type{font-weight:bold;text-decoration:none;margin-left:-1.05em}
#footnotes .footnote:last-of-type{margin-bottom:0}
#content #footnotes{margin-top:-.625em;margin-bottom:0;padding:.75em 0}
.gist .file-data>table{border:0;background:#fff;width:100%;margin-bottom:0}
.gist .file-data>table td.line-data{width:99%}
div.unbreakable{page-break-inside:avoid}
.big{font-size:larger}
.small{font-size:smaller}
.underline{text-decoration:underline}
.overline{text-decoration:overline}
.line-through{text-decoration:line-through}
.aqua{color:#00bfbf}
.aqua-background{background:#00fafa}
.black{color:#000}
.black-background{background:#000}
.blue{color:#0000bf}
.blue-background{background:#0000fa}
.fuchsia{color:#bf00bf}
.fuchsia-background{background:#fa00fa}
.gray{color:#606060}
.gray-background{background:#7d7d7d}
.green{color:#006000}
.green-background{background:#007d00}
.lime{color:#00bf00}
.lime-background{background:#00fa00}
.maroon{color:#600000}
.maroon-background{background:#7d0000}
.navy{color:#000060}
.navy-background{background:#00007d}
.olive{color:#606000}
.olive-background{background:#7d7d00}
.purple{color:#600060}
.purple-background{background:#7d007d}
.red{color:#bf0000}
.red-background{background:#fa0000}
.silver{color:#909090}
.silver-background{background:#bcbcbc}
.teal{color:#006060}
.teal-background{background:#007d7d}
.white{color:#bfbfbf}
.white-background{background:#fafafa}
.yellow{color:#bfbf00}
.yellow-background{background:#fafa00}
span.icon>.fa{cursor:default}
a span.icon>.fa{cursor:inherit}
.admonitionblock td.icon [class^="fa icon-"]{font-size:2.5em;text-shadow:1px 1px 2px rgba(0,0,0,.5);cursor:default}
.admonitionblock td.icon .icon-note::before{content:"\f05a";color:#19407c}
.admonitionblock td.icon .icon-tip::before{content:"\f0eb";text-shadow:1px 1px 2px rgba(155,155,0,.8);color:#111}
.admonitionblock td.icon .icon-warning::before{content:"\f071";color:#bf6900}
.admonitionblock td.icon .icon-caution::before{content:"\f06d";color:#bf3400}
.admonitionblock td.icon .icon-important::before{content:"\f06a";color:#bf0000}
.conum[data-value]{display:inline-block;color:#fff!important;background:rgba(0,0,0,.8);-webkit-border-radius:50%;border-radius:50%;text-align:center;font-size:.75em;width:1.67em;height:1.67em;line-height:1.67em;font-family:"Open Sans","DejaVu Sans",sans-serif;font-style:normal;font-weight:bold}
.conum[data-value] *{color:#fff!important}
.conum[data-value]+b{display:none}
.conum[data-value]::after{content:attr(data-value)}
pre .conum[data-value]{position:relative;top:-.125em}
b.conum *{color:inherit!important}
.conum:not([data-value]):empty{display:none}
dt,th.tableblock,td.content,div.footnote{text-rendering:optimizeLegibility}
h1,h2,p,td.content,span.alt{letter-spacing:-.01em}
p strong,td.content strong,div.footnote strong{letter-spacing:-.005em}
p,blockquote,dt,td.content,span.alt{font-size:1.0625rem}
p{margin-bottom:1.25rem}
.sidebarblock p,.sidebarblock dt,.sidebarblock td.content,p.tableblock{font-size:1em}
.exampleblock>.content{background:#fffef7;border-color:#e0e0dc;-webkit-box-shadow:0 1px 4px #e0e0dc;box-shadow:0 1px 4px #e0e0dc}
.print-only{display:none!important}
@page{margin:1.25cm .75cm}
@media print{*{-webkit-box-shadow:none!important;box-shadow:none!important;text-shadow:none!important}
html{font-size:80%}
a{color:inherit!important;text-decoration:underline!important}
a.bare,a[href^="#"],a[href^="mailto:"]{text-decoration:none!important}
a[href^="http:"]:not(.bare)::after,a[href^="https:"]:not(.bare)::after{content:"(" attr(href) ")";display:inline-block;font-size:.875em;padding-left:.25em}
abbr[title]::after{content:" (" attr(title) ")"}
pre,blockquote,tr,img,object,svg{page-break-inside:avoid}
thead{display:table-header-group}
svg{max-width:100%}
p,blockquote,dt,td.content{font-size:1em;orphans:3;widows:3}
h2,h3,#toctitle,.sidebarblock>.content>.title{page-break-after:avoid}
#header,#content,#footnotes,#footer{max-width:none}
#toc,.sidebarblock,.exampleblock>.content{background:none!important}
#toc{border-bottom:1px solid #dddddf!important;padding-bottom:0!important}
body.book #header{text-align:center}
body.book #header>h1:first-child{border:0!important;margin:2.5em 0 1em}
body.book #header .details{border:0!important;display:block;padding:0!important}
body.book #header .details span:first-child{margin-left:0!important}
body.book #header .details br{display:block}
body.book #header .details br+span::before{content:none!important}
body.book #toc{border:0!important;text-align:left!important;padding:0!important;margin:0!important}
body.book #toc,body.book #preamble,body.book h1.sect0,body.book .sect1>h2{page-break-before:always}
.listingblock code[data-lang]::before{display:block}
#footer{padding:0 .9375em}
.hide-on-print{display:none!important}
.print-only{display:block!important}
.hide-for-print{display:none!important}
.show-for-print{display:inherit!important}}
@media print,amzn-kf8{#header>h1:first-child{margin-top:1.25rem}
.sect1{padding:0!important}
.sect1+.sect1{border:0}
#footer{background:none}
#footer-text{color:rgba(0,0,0,.6);font-size:.9em}}
@media amzn-kf8{#header,#content,#footnotes,#footer{padding:0}}
</style>
<style>
pre { line-height: 125%; }
td.linenos pre { color: #000000; background-color: #f0f0f0; padding-left: 5px; padding-right: 5px; }
span.linenos { color: #000000; background-color: #f0f0f0; padding-left: 5px; padding-right: 5px; }
td.linenos pre.special { color: #000000; background-color: #ffffc0; padding-left: 5px; padding-right: 5px; }
span.linenos.special { color: #000000; background-color: #ffffc0; padding-left: 5px; padding-right: 5px; }
pre.pygments .hll { background-color: #ffffcc }
pre.pygments { background: #f8f8f8; }
pre.pygments .tok-c { color: #008800; font-style: italic } /* Comment */
pre.pygments .tok-err { border: 1px solid #FF0000 } /* Error */
pre.pygments .tok-k { color: #AA22FF; font-weight: bold } /* Keyword */
pre.pygments .tok-o { color: #666666 } /* Operator */
pre.pygments .tok-ch { color: #008800; font-style: italic } /* Comment.Hashbang */
pre.pygments .tok-cm { color: #008800; font-style: italic } /* Comment.Multiline */
pre.pygments .tok-cp { color: #008800 } /* Comment.Preproc */
pre.pygments .tok-cpf { color: #008800; font-style: italic } /* Comment.PreprocFile */
pre.pygments .tok-c1 { color: #008800; font-style: italic } /* Comment.Single */
pre.pygments .tok-cs { color: #008800; font-weight: bold } /* Comment.Special */
pre.pygments .tok-gd { color: #A00000 } /* Generic.Deleted */
pre.pygments .tok-ge { font-style: italic } /* Generic.Emph */
pre.pygments .tok-gr { color: #FF0000 } /* Generic.Error */
pre.pygments .tok-gh { color: #000080; font-weight: bold } /* Generic.Heading */
pre.pygments .tok-gi { color: #00A000 } /* Generic.Inserted */
pre.pygments .tok-go { color: #888888 } /* Generic.Output */
pre.pygments .tok-gp { color: #000080; font-weight: bold } /* Generic.Prompt */
pre.pygments .tok-gs { font-weight: bold } /* Generic.Strong */
pre.pygments .tok-gu { color: #800080; font-weight: bold } /* Generic.Subheading */
pre.pygments .tok-gt { color: #0044DD } /* Generic.Traceback */
pre.pygments .tok-kc { color: #AA22FF; font-weight: bold } /* Keyword.Constant */
pre.pygments .tok-kd { color: #AA22FF; font-weight: bold } /* Keyword.Declaration */
pre.pygments .tok-kn { color: #AA22FF; font-weight: bold } /* Keyword.Namespace */
pre.pygments .tok-kp { color: #AA22FF } /* Keyword.Pseudo */
pre.pygments .tok-kr { color: #AA22FF; font-weight: bold } /* Keyword.Reserved */
pre.pygments .tok-kt { color: #00BB00; font-weight: bold } /* Keyword.Type */
pre.pygments .tok-m { color: #666666 } /* Literal.Number */
pre.pygments .tok-s { color: #BB4444 } /* Literal.String */
pre.pygments .tok-na { color: #BB4444 } /* Name.Attribute */
pre.pygments .tok-nb { color: #AA22FF } /* Name.Builtin */
pre.pygments .tok-nc { color: #0000FF } /* Name.Class */
pre.pygments .tok-no { color: #880000 } /* Name.Constant */
pre.pygments .tok-nd { color: #AA22FF } /* Name.Decorator */
pre.pygments .tok-ni { color: #999999; font-weight: bold } /* Name.Entity */
pre.pygments .tok-ne { color: #D2413A; font-weight: bold } /* Name.Exception */
pre.pygments .tok-nf { color: #00A000 } /* Name.Function */
pre.pygments .tok-nl { color: #A0A000 } /* Name.Label */
pre.pygments .tok-nn { color: #0000FF; font-weight: bold } /* Name.Namespace */
pre.pygments .tok-nt { color: #008000; font-weight: bold } /* Name.Tag */
pre.pygments .tok-nv { color: #B8860B } /* Name.Variable */
pre.pygments .tok-ow { color: #AA22FF; font-weight: bold } /* Operator.Word */
pre.pygments .tok-w { color: #bbbbbb } /* Text.Whitespace */
pre.pygments .tok-mb { color: #666666 } /* Literal.Number.Bin */
pre.pygments .tok-mf { color: #666666 } /* Literal.Number.Float */
pre.pygments .tok-mh { color: #666666 } /* Literal.Number.Hex */
pre.pygments .tok-mi { color: #666666 } /* Literal.Number.Integer */
pre.pygments .tok-mo { color: #666666 } /* Literal.Number.Oct */
pre.pygments .tok-sa { color: #BB4444 } /* Literal.String.Affix */
pre.pygments .tok-sb { color: #BB4444 } /* Literal.String.Backtick */
pre.pygments .tok-sc { color: #BB4444 } /* Literal.String.Char */
pre.pygments .tok-dl { color: #BB4444 } /* Literal.String.Delimiter */
pre.pygments .tok-sd { color: #BB4444; font-style: italic } /* Literal.String.Doc */
pre.pygments .tok-s2 { color: #BB4444 } /* Literal.String.Double */
pre.pygments .tok-se { color: #BB6622; font-weight: bold } /* Literal.String.Escape */
pre.pygments .tok-sh { color: #BB4444 } /* Literal.String.Heredoc */
pre.pygments .tok-si { color: #BB6688; font-weight: bold } /* Literal.String.Interpol */
pre.pygments .tok-sx { color: #008000 } /* Literal.String.Other */
pre.pygments .tok-sr { color: #BB6688 } /* Literal.String.Regex */
pre.pygments .tok-s1 { color: #BB4444 } /* Literal.String.Single */
pre.pygments .tok-ss { color: #B8860B } /* Literal.String.Symbol */
pre.pygments .tok-bp { color: #AA22FF } /* Name.Builtin.Pseudo */
pre.pygments .tok-fm { color: #00A000 } /* Name.Function.Magic */
pre.pygments .tok-vc { color: #B8860B } /* Name.Variable.Class */
pre.pygments .tok-vg { color: #B8860B } /* Name.Variable.Global */
pre.pygments .tok-vi { color: #B8860B } /* Name.Variable.Instance */
pre.pygments .tok-vm { color: #B8860B } /* Name.Variable.Magic */
pre.pygments .tok-il { color: #666666 } /* Literal.Number.Integer.Long */
</style>
</head>
<body class="book toc2 toc-left">
<div id="header">
<h1>Flu-crew Technical Documentation</h1>
<div id="toc" class="toc2">
<div id="toctitle">Table of Contents</div>
<ul class="sectlevel0">
<li><a href="#_our_data_and_how_to_manage_it">Our data and how to manage it</a>
<ul class="sectlevel1">
<li><a href="#_the_octoflu_database">1. The OctoFLU Database</a>
<ul class="sectlevel2">
<li><a href="#_installation">1.1. Installation</a></li>
<li><a href="#_database_management_with_the_octofludb_utility">1.2. Database management with the <code>octofludb</code> utility</a></li>
<li><a href="#_getting_data">1.3. Getting data</a></li>
<li><a href="#_deleting_data">1.4. Deleting data</a>
<ul class="sectlevel3">
<li><a href="#_delete_one_link">1.4.1. delete one link</a></li>
<li><a href="#_delete_many_links">1.4.2. delete many links</a></li>
</ul>
</li>
<li><a href="#_uploading_data">1.5. Uploading data</a></li>
<li><a href="#_recipes">1.6. recipes</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="#_methods_for_sampling_and_coloring_trees">Methods for sampling and coloring trees</a>
<ul class="sectlevel1">
<li><a href="#_methods_for_manipulating_trees">2. Methods for manipulating trees</a>
<ul class="sectlevel2">
<li><a href="#_pandemic_h1_filtering_down_a_giant_clade">2.1. Pandemic H1 - filtering down a giant clade</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="#_methods_for_working_with_sequence_data">Methods for working with sequence data</a>
<ul class="sectlevel1">
<li><a href="#_smof_simple_manipulation_of_fasta">3. smof - Simple Manipulation Of Fasta</a></li>
<li><a href="#_motif_extraction">4. Motif extraction</a></li>
<li><a href="#_flutile_specialized_tools_for_influenza">5. <code>flutile</code> - specialized tools for influenza</a>
<ul class="sectlevel2">
<li><a href="#_ha1_extraction">5.1. HA1 extraction</a></li>
<li><a href="#_flutile_and_aadiff_table_generation">5.2. <code>flutile</code> and aadiff table generation</a></li>
<li><a href="#_flutile_and_indexing">5.3. <code>flutile</code> and indexing</a></li>
<li><a href="#_flutile_and_motif_extraction">5.4. <code>flutile</code> and motif extraction</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="#_data_visualization">Data Visualization</a>
<ul class="sectlevel1">
<li><a href="#_octoflushow">6. octoflushow</a></li>
</ul>
</li>
<li><a href="#_clade_inference">Clade inference</a>
<ul class="sectlevel1">
<li><a href="#_octoflu">7. octoFLU</a></li>
</ul>
</li>
<li><a href="#_standard_reports">Standard Reports</a>
<ul class="sectlevel1">
<li><a href="#_quarterly_report">8. quarterly report</a></li>
</ul>
</li>
<li><a href="#_whole_genome_sequencing_monthly_selection">Whole Genome Sequencing monthly selection</a>
<ul class="sectlevel1">
<li><a href="#_offlu">9. offlu</a></li>
</ul>
</li>
<li><a href="#_literature">Literature</a>
<ul class="sectlevel1">
<li><a href="#_biology_papers">10. Biology papers</a></li>
<li><a href="#_method_papers">11. Method papers</a></li>
</ul>
</li>
<li><a href="#_glossary">Glossary</a></li>
<li><a href="#_references">References</a></li>
</ul>
</div>
</div>
<div id="content">
<h1 id="_our_data_and_how_to_manage_it" class="sect0">Our data and how to manage it</h1>
<div class="sect1">
<h2 id="_the_octoflu_database">1. The OctoFLU Database</h2>
<div class="sectionbody">
<div class="paragraph">
<p>The OctoFLU Database is a platform for storing, curating, and integrating all
our data.  It is built on the GraphDB database (<a href="http://graphdb.ontotext.com/" class="bare">http://graphdb.ontotext.com/</a>).</p>
</div>
<div class="paragraph">
<p>The database can be accessed either through the graphical user interface or the
<code>octofludb</code> command line utility. The latter approach is strongly recommended
and will be the only approach I discuss in the future sections.</p>
</div>
<div class="sect2">
<h3 id="_installation">1.1. Installation</h3>
<div class="paragraph">
<p>The free version of GraphDB can be downloaded and installed from the website.</p>
</div>
<div class="paragraph">
<p><code>octofludb</code> can be installed through pip:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>pip install octofludb</pre>
</div>
</div>
<div class="paragraph">
<p>You may need to replace <code>pip</code> with <code>pip3</code>, depending on your setup.</p>
</div>
</div>
<div class="sect2">
<h3 id="_database_management_with_the_octofludb_utility">1.2. Database management with the <code>octofludb</code> utility</h3>
<div class="paragraph">
<p>Documentation for all subcommands in <code>octofludb</code> can be accessed by calling
<code>octofludb -h</code>.  Subcommand documentation can be accessed with <code>octofludb
&lt;subcommand&gt; -h</code>.</p>
</div>
<div class="paragraph">
<p>Once GraphDB is installed on your system, you can initialize an empty repo with
<code>octofludb init</code>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_getting_data">1.3. Getting data</h3>
<div class="paragraph">
<p>A local database can be accessed by submitting SPARQL queries. A SPARQL script
specifies a pattern over a graph and when given to the database will print a
row of data for every occurrence of the pattern.</p>
</div>
<div class="paragraph">
<p>Below is a very simple script that returns the strain names of every gamma:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="pygments highlight"><code data-lang="sparql"><span></span><span class="tok-k">PREFIX</span> <span class="tok-nn">f</span><span class="tok-p">:</span> <span class="tok-nl">&lt;https://flu-crew.org/term/&gt;</span>

<span class="tok-k">SELECT</span> <span class="tok-nv">?name</span>
<span class="tok-k">WHERE</span> <span class="tok-p">{</span>
  <span class="tok-nv">?sid</span> <span class="tok-nn">f</span><span class="tok-p">:</span><span class="tok-nt">strain_name</span> <span class="tok-nv">?name</span> <span class="tok-p">.</span>
  <span class="tok-nv">?sid</span> <span class="tok-nn">f</span><span class="tok-p">:</span><span class="tok-nt">host</span> <span class="tok-s">&quot;swine&quot;</span> <span class="tok-p">.</span>
  <span class="tok-nv">?sid</span> <span class="tok-nn">f</span><span class="tok-p">:</span><span class="tok-nt">has_segment</span> <span class="tok-nv">?gid</span> <span class="tok-p">.</span>
  <span class="tok-nv">?gid</span> <span class="tok-nn">f</span><span class="tok-p">:</span><span class="tok-nt">clade</span> <span class="tok-s">&quot;gamma&quot;</span> <span class="tok-p">.</span>
<span class="tok-p">}</span></code></pre>
</div>
</div>
<div class="paragraph">
<p>The PREFIX line defines an abbreviatin, <code>f</code>, for the long URI
<code><a href="https://flu-crew.org/term/" class="bare">https://flu-crew.org/term/</a></code>. While the URI looks like a web link, it does not
need to point to a real web page (although it may, and that is a good place to
put documentation). Every edge type defined specifically for OctofluDB has this
prefix.</p>
</div>
<div class="paragraph">
<p>The SELECT term is followed by a list of all the columns we want printed out
for us, in this case just one.</p>
</div>
<div class="paragraph">
<p><code>?sid f:strain_name ?name .</code> is a graph pattern restricting the returned data
to things related to eachother by the <code>f:strain_name</code> predicate. Words that
begin with <code>?</code> are variables. There is nothing special about <code>?sid</code> and
<code>?name</code>, these variables could be named anything. By convention, I normally use
<code>?sid</code> for a *s*train *id*entifier and <code>?gid</code> for a *g*enbank *id*enditifer.
Every statement must end with a <code>.</code>, otherwise GraphDB will spill out a very
ugly error statement.</p>
</div>
<div class="paragraph">
<p><code>?sid f:has_segment ?gid .</code> links our strain node to its child segment node.
This particular link will appear in many queries.</p>
</div>
<div class="paragraph">
<p><code>?gid f:clade "gamma"</code> links the segment to the literal string value "gamma" by the <code>f:clade</code> relation.</p>
</div>
<div class="paragraph">
<p>The query can be sent to the database with the command:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>octofludb query fetch-gammas.rq</pre>
</div>
</div>
<div class="paragraph">
<p>The extension <code>.rq</code> is convention for SPARQL scripts.</p>
</div>
</div>
<div class="sect2">
<h3 id="_deleting_data">1.4. Deleting data</h3>
<div class="paragraph">
<p>I will describe how to delete data with a few examples.</p>
</div>
<div class="sect3">
<h4 id="_delete_one_link">1.4.1. delete one link</h4>
<div class="paragraph">
<p>To understand the need for deleting data, you should know a little about how
graph databases handle conflicting data. Data uploaded to GraphDB consists of
many "triples", where a triple consists of a subject, a predicate
(relationship), and an object. That is, a triple specifies a single edge in the
graph.</p>
</div>
<div class="paragraph">
<p>For example, suppose we have an edge in the database linking the HA segment
EPI_123456 to the clade "gamma". Then suppose we want to instead link it to the
more specific clade "gamma.3". If we upload the new triple, however, the
original link "gamma" is not removed, instead we end up with the segment being
linked to both "gamma" and "gamma.3". The database doesn&#8217;t magically know that
the relation "clade" should always be one-to-one. So if we want only the clade
"gamma.3", we will need to delete the original clade.</p>
</div>
<div class="paragraph">
<p>The following SPARQL command will do what we want:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="pygments highlight"><code data-lang="sparql"><span></span><span class="tok-k">PREFIX</span> <span class="tok-nn">f</span><span class="tok-p">:</span> <span class="tok-nl">&lt;https://flu-crew.org/term/&gt;</span>

<span class="tok-k">DELETE</span> <span class="tok-p">{</span>
    <span class="tok-nv">?gid</span> <span class="tok-nn">f</span><span class="tok-p">:</span><span class="tok-nt">epi_id</span> <span class="tok-s">&quot;EPI_123456&quot;</span> <span class="tok-p">.</span>
    <span class="tok-nv">?gid</span> <span class="tok-nn">f</span><span class="tok-p">:</span><span class="tok-nt">clade</span> <span class="tok-s">&quot;gamma&quot;</span> <span class="tok-p">.</span>
<span class="tok-p">}</span></code></pre>
</div>
</div>
<div class="paragraph">
<p>This deletion script can be written to a file, <code>delete-clade.rq</code>, and executed
on the local database with <code>octofludb update delete-clade.rq</code>.</p>
</div>
</div>
<div class="sect3">
<h4 id="_delete_many_links">1.4.2. delete many links</h4>
<div class="paragraph">
<p>This example expands on the prior example and introduces a more powerful
technique for deleting data.</p>
</div>
<div class="paragraph">
<p>In the August OFFLU report process, we updated the octoFLU reference files to
better represent the global H3 clades. If we directly upload the new H3 clades
to octofludb, then we will end up with duplicate records. So we must first
delete all prior non-USA H3 clades. The SPARQL deletion statement for this is
shown below.</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="pygments highlight"><code data-lang="sparql"><span></span><span class="tok-k">PREFIX</span> <span class="tok-nn">f</span><span class="tok-p">:</span> <span class="tok-nl">&lt;https://flu-crew.org/term/&gt;</span>

<span class="tok-k">DELETE</span> <span class="tok-p">{</span>
    <span class="tok-nv">?gid</span> <span class="tok-nn">f</span><span class="tok-p">:</span><span class="tok-nt">clade</span> <span class="tok-nv">?clade</span> <span class="tok-p">.</span>
    <span class="tok-nv">?gid</span> <span class="tok-nn">f</span><span class="tok-p">:</span><span class="tok-nt">gl_clade</span> <span class="tok-nv">?gl_clade</span> <span class="tok-p">.</span>
<span class="tok-p">}</span>
<span class="tok-k">WHERE</span> <span class="tok-p">{</span>
  <span class="tok-nv">?sid</span> <span class="tok-nn">f</span><span class="tok-p">:</span><span class="tok-nt">strain_name</span> <span class="tok-nv">?strain_name</span> <span class="tok-p">.</span>
  <span class="tok-nv">?sid</span> <span class="tok-nn">f</span><span class="tok-p">:</span><span class="tok-nt">country</span><span class="tok-o">/</span><span class="tok-nn">f</span><span class="tok-p">:</span><span class="tok-nt">code</span> <span class="tok-nv">?country</span> <span class="tok-p">.</span>
  <span class="tok-nv">?sid</span> <span class="tok-nn">f</span><span class="tok-p">:</span><span class="tok-nt">host</span> <span class="tok-s">&quot;swine&quot;</span> <span class="tok-p">.</span>

  <span class="tok-k">FILTER</span> <span class="tok-p">(</span><span class="tok-nv">?country</span> <span class="tok-o">!=</span> <span class="tok-s">&quot;USA&quot;</span><span class="tok-p">)</span> <span class="tok-p">.</span>

  <span class="tok-nv">?sid</span> <span class="tok-nn">f</span><span class="tok-p">:</span><span class="tok-nt">subtype</span> <span class="tok-nv">?subtype</span> <span class="tok-p">.</span>
  <span class="tok-k">FILTER</span> <span class="tok-nf">REGEX</span><span class="tok-p">(</span><span class="tok-nv">?subtype</span><span class="tok-p">,</span> <span class="tok-s">&quot;H3&quot;</span><span class="tok-p">)</span> <span class="tok-p">.</span>

  <span class="tok-nv">?sid</span> <span class="tok-nn">f</span><span class="tok-p">:</span><span class="tok-nt">has_segment</span> <span class="tok-nv">?gid</span> <span class="tok-p">.</span>
  <span class="tok-nv">?gid</span> <span class="tok-nn">f</span><span class="tok-p">:</span><span class="tok-nt">segment_name</span> <span class="tok-s">&quot;HA&quot;</span> <span class="tok-p">.</span>

  <span class="tok-k">OPTIONAL</span> <span class="tok-p">{</span> <span class="tok-nv">?gid</span> <span class="tok-nn">f</span><span class="tok-p">:</span><span class="tok-nt">clade</span> <span class="tok-nv">?clade</span> <span class="tok-p">.</span> <span class="tok-p">}</span>
  <span class="tok-k">OPTIONAL</span> <span class="tok-p">{</span> <span class="tok-nv">?gid</span> <span class="tok-nn">f</span><span class="tok-p">:</span><span class="tok-nt">gl_clade</span> <span class="tok-nv">?gl_clade</span> <span class="tok-p">.</span> <span class="tok-p">}</span>
<span class="tok-p">}</span></code></pre>
</div>
</div>
<div class="paragraph">
<p>The WHERE clause defines a set of triples and the DELETE clause specifies which
of these triples should be removed.</p>
</div>
<div class="paragraph">
<p>This command should be written to a file (say, <code>delete-global-H3-clades.rq</code>)
and then can be applied to the local database with the command:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>octofludb update delete-global-H3-clades.rq</pre>
</div>
</div>
<div class="paragraph">
<p>It is generally good practice to check before we delete. For example, if we
accidentally added the triple <code>?sid f:has_segment ?gid</code> to the DELETE clause,
then we would delete the links from strains to segments. An easy way to check
what will be deleted is to create the corresponding query statement:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="pygments highlight"><code data-lang="sparql"><span></span><span class="tok-k">PREFIX</span> <span class="tok-nn">f</span><span class="tok-p">:</span> <span class="tok-nl">&lt;https://flu-crew.org/term/&gt;</span>

<span class="tok-k">SELECT</span> <span class="tok-o">*</span>
<span class="tok-k">WHERE</span> <span class="tok-p">{</span>
   <span class="tok-p">....</span>
<span class="tok-p">}</span></code></pre>
</div>
</div>
<div class="paragraph">
<p>With the WHERE clause being the same as before. The calling <code>octofludb query</code>
to get a table showing all cases where clade info will be deleted.</p>
</div>
</div>
</div>
<div class="sect2">
<h3 id="_uploading_data">1.5. Uploading data</h3>
<div class="paragraph">
<p>There are three basic steps to uploading data to the database.</p>
</div>
<div class="paragraph">
<p>The first is to prepare the raw data in the "proper" way.</p>
</div>
<div class="paragraph">
<p>The second step is to convert the raw data to the turtle-format GraphDB
understands. This is done through the <code>mk_*</code> commands in <code>octofludb</code>. These
commands do more than just translate the data, they use specialized parsers to
extract flu specific information from the data. For example, if</p>
</div>
<div class="paragraph">
<p>The third step is to upload the turtle files to the database. If you are
working with a local database, this step is easy: <code>octofludb upload
&lt;turtle-fiels&gt;</code>.</p>
</div>
</div>
<div class="sect2">
<h3 id="_recipes">1.6. recipes</h3>
<div class="paragraph">
<p>Recipe functions are data synthesis functions that are built into <code>octofludb</code>.
They will be added on an as-needed database.</p>
</div>
<div class="ulist">
<ul>
<li>
<p><code>masterlist</code> - generate a table of data describing all A0 strains. This is
the base table used by <code>octoflushow</code> and to generate quarterly reports.</p>
</li>
</ul>
</div>
</div>
</div>
</div>
<h1 id="_methods_for_sampling_and_coloring_trees" class="sect0">Methods for sampling and coloring trees</h1>
<div class="sect1">
<h2 id="_methods_for_manipulating_trees">2. Methods for manipulating trees</h2>
<div class="sectionbody">
<div class="paragraph">
<p>I will describe methods here through a series of examples.</p>
</div>
<div class="sect2">
<h3 id="_pandemic_h1_filtering_down_a_giant_clade">2.1. Pandemic H1 - filtering down a giant clade</h3>
<div class="paragraph">
<p>You have just retrieved H1 pandemic sequence from the last few years and want
to do something awesome with it. With <code>smof</code>, you can see how many sequences
you have:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>$ smof stat pdm.fna
nseq:      26802
nchars:    26257059
5sum:      1 982 982 982 991
mean(sd):  980 (29)
N50:       982</pre>
</div>
</div>
<div class="paragraph">
<p>Oh crap, one of our sequences has a length of 1.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>$ smof filter -s 100 pdm.fna
&gt;|A/Jacarei/1495954-IAL/2020|H1N1|human|BRA|||2020-02-29
a
&gt;|A/Presidente_Venceslau/1261099-IAL/2019|H1N1|human|BRA|||2019-08-13
a
&gt;|A/Goiania/1435-IAL/2019|H1N1|human|BRA|||2019-09-07
c
&gt;|A/Sao_Paulo/167/2016|H1N1|human|BRA|||2016-01-04
g
&gt;|A/Sao_Paulo/725017/2018|H1N1|human|BRA|||2018-02-27
g</pre>
</div>
</div>
<div class="paragraph">
<p>The HAs were trimmed to the HA1 region, so these sequences are simply truncated
sequences. We can either filter them out or change our methods to not trim.
I&#8217;ll go with the third option of leaving them in and building the tree (since
this is the tree section).</p>
</div>
<div class="listingblock">
<div class="content">
<pre>$ mafft pdm.fna | FastTree -nt -gamma -gtr &gt; pdm.tre</pre>
</div>
</div>
<div class="paragraph">
<p>This will take awhile to run, since there are nearly 27 thousand strains. The
result tree is shown <a href="#fig:tree1">below</a>.</p>
</div>
<div id="fig:tree1" class="paragraph">
<p><span class="image"><img src="images/tree1.png" alt="tree1"></span></p>
</div>
<div class="paragraph">
<p>We want to root it on California/07/2009 and then sort it. This can be done
with the <code>treebender</code> utility from the phylommand package
(<a href="https://github.com/RybergGroup/phylommand" class="bare">https://github.com/RybergGroup/phylommand</a>):</p>
</div>
<div class="listingblock">
<div class="content">
<pre>$ treebender --outgroup_root "vaccine|A/California/07/2009|human" pdm.tre | treebender --ladderize &gt; pdm2.tre</pre>
</div>
</div>
<div class="paragraph">
<p>The outgroup name needs to be an exact match against the tip label. The result is much cleaner:</p>
</div>
<div id="fig:tree2" class="paragraph">
<p><span class="image"><img src="images/tree2.png" alt="tree2"></span></p>
</div>
<div class="paragraph">
<p>But this tree is far too large for our purposes. We need to subsample. However,
since we are interested in human-swine transmission, we need to preserve the
overall topology of the tree with respect to the two species. If we just
sampled 5% of the tips, we would loose most of the non-persisting spill-over
cases. <code>smot</code> can sample for us while preserving the topology.</p>
</div>
<div class="listingblock">
<div class="content">
<pre>smot sample para --keep "swine" --seed 42 -p 0.02 --min-tips=3 --factor-by-capture="(human|swine)" pdm2.tre &gt; pdm3.tre</pre>
</div>
</div>
<div class="paragraph">
<p>This yields a much more manageable tree with only 2643 tips. The black and
orange represent human and swine sequences, respectively.</p>
</div>
<div id="fig:tree3" class="paragraph">
<p><span class="image"><img src="images/tree3.png" alt="tree3"></span></p>
</div>
<div class="paragraph">
<p>If you just want to see the swine, <code>smot</code> can grep them out:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>smot grep "swine" pdm3.tre &gt; pdm-swine.tre</pre>
</div>
</div>
<div class="paragraph">
<p>Or if you want to remove the private VDL sequences:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>smot grep -v "VDL" pdm3.tre &gt; pdm-no-vdl.tre</pre>
</div>
</div>
</div>
</div>
</div>
<h1 id="_methods_for_working_with_sequence_data" class="sect0">Methods for working with sequence data</h1>
<div class="openblock partintro">
<div class="content">
Methods for manipulating FASTA files
</div>
</div>
<div class="sect1">
<h2 id="_smof_simple_manipulation_of_fasta">3. smof - Simple Manipulation Of Fasta</h2>
<div class="sectionbody">
<div class="paragraph">
<p><code>smof</code> is a broad set of CLI tools for working with fasta files. It can be
installed through <code>pip</code>:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="pygments highlight"><code data-lang="bash"><span></span>pip install smof</code></pre>
</div>
</div>
<div class="paragraph">
<p>If you are on a MacOS, you may need to use <code>pip3</code> instead. <code>smof</code> works on
Linux, MacOS, and Windows. The only dependency is Python v3.4 or greater.</p>
</div>
<div class="paragraph">
<p>A summary of the tools it includes can be accessed by calling <code>smof -h</code> in the
terminal.  Here is the usage statement (as of v0.19.0):</p>
</div>
<div class="listingblock">
<div class="content">
<pre>usage: &lt;fastafile&gt; | smof &lt;subcommand&gt; &lt;options&gt;

Tools for studying and manipulating fasta files

optional arguments:
  -h, --help            show this help message and exit
  -v, --version         show program's version number and exit

subcommands:
  [ for help on each: smof &lt;subcommand&gt; -h ]
    cut                 emulates UNIX cut command, where fields are entries
    clean               cleans fasta files
    consensus           finds the consensus sequence for aligned sequence
    filter              extracts sequences meeting the given conditions
    grep                roughly emulates the UNIX grep command
    md5sum              calculate an md5 checksum for the input sequences
    head                writes the first sequences in a file
    permute             randomly order sequence
    reverse             reverse each sequence (or reverse complement)
    sample              randomly select entries from fasta file
    sniff               extract info about the sequence
    sort                sort sequences
    split               split a fasta file into smaller files
    stat                calculate sequence statistics
    subseq              extract subsequence from each entry (revcomp if a&lt;b)
    tail                writes the last sequences in a file
    translate           translate a DNA sequence into a protein sequence
    uniq                count, omit, or merge repeated entries
    wc                  roughly emulates the UNIX wc command

Project site: https://github.com/incertae-sedis/smof
Report bugs/requests via https://github.com/incertae-sedis/smof/issues
Author: Zebulun Arendsee (zbwrnz@gmail.com)</pre>
</div>
</div>
<div class="paragraph">
<p>Detailed information about each subcommand can be accessed with <code>smof
&lt;subcommand&gt; -h</code>, for example:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>$ smof stat -h

usage: &lt;fastafile&gt; | smof stat &lt;options&gt;

The default action is to count the lengths of all sequences and output summary
statistics including: 1) the number of sequences, 2) the number of characters,
3) the five-number summary of sequence lengths (minimum, 25th quantile, median,
75th quantile, and maximum), 4) the mean and standard deviation of lengths, and
5) the N50 (if you don't know what that is, you don't need to know).

positional arguments:
  INPUT                 input fasta sequence (default = stdin)

optional arguments:
  -h, --help            show this help message and exit
  -d DELIMITER, --delimiter DELIMITER
                        output delimiter
  -q, --byseq           write a line for each sequence
  -I, --case-sensitive  match case
  -m, --count-lower     count the number of lowercase characters
  -c, --counts          write counts of all characters
  -t, --type            guess sequence type
  -l, --length          write sequence length
  -p, --proportion      write proportion of each character
  -C, --aa-profile      display protein profile
  -g, --hist            write ascii histogram of sequence lengths
  -G, --log-hist        write ascii histogram of sequence log2 lengths</pre>
</div>
</div>
<div class="paragraph">
<p>For more information on <code>smof</code>, visit the repo at url:[<a href="https://github.com/incertae-sedis/smof" class="bare">https://github.com/incertae-sedis/smof</a>].</p>
</div>
<div class="paragraph">
<p>While <code>smof</code> may have an overwelming number of features, there are just a few
that you will be likely to use in day-to-day influenza analysis. I will
introduce just a few below.</p>
</div>
<div class="olist arabic">
<ol class="arabic">
<li>
<p><code>smof stat</code> - I use this all the time to just count the sequences in a
fasta file and check for oddly-sized entries.</p>
</li>
<li>
<p><code>smof sniff</code> - The output for this one is a bit cryptic (see <code>smof sniff
-h</code>), but it will tell you if there are repeated sequences, how many of the
sequences appear to be DNA or protein, whether there are internal stops in a
protein sequence, and whether any sequences contain illegal characters. It is
a good idea to run a FASTA file through sniff just to be sure went South in
your (or someone elses) scripting.</p>
</li>
<li>
<p><code>smof translate -f</code> - Translate the longest coding frame. This works
perfectly for translating flu segments to proteins. Before using <code>smof</code> for
this purpose, we had to manually align sequences in Geneious, trim them, and
then translate them.</p>
<div class="ulist">
<ul>
<li>
<p><code>-c</code> - add this flag want the untranslated codon sequence</p>
</li>
<li>
<p><code>-s</code> - add this flag if you want the translation to begin with a START codon</p>
</li>
</ul>
</div>
</li>
<li>
<p><code>smof uniq -f</code> - If any entries have duplicate headers, keep only the
first. Some programs, such as FastTree, will die if there are any duplicate
headers. This command is a fast-and-dirty way to remove these duplicates.</p>
</li>
<li>
<p><code>smof uniq -p -z "<em><em>"</code> - Combine entries with identical sequence by
concatenating the headers. I use '</em></em>' as the header separater since it is
easy to see and unlikely to appear in a normal header. Often I will grep out
the combined headers grep '___' myfile.fasta to see what has been combined.
Then I may either remove the duplicate header info (<code>sed 's/<em>_</em>.*//'</code>) or open
the fasta in an editor and manually choose what information to include.</p>
</li>
<li>
<p><code>smof clean -ux</code> - Clean up a fasta file, remove any inserted gaps, wrap at
80 characters, convert to uppercase, remove any Windows carriage returns. For
DNA sequence, I usually run <code>smof clean -t n -uxdr</code>. The <code>-t n</code> tells <code>smof</code>
that the sequence is nucleotides, the <code>-d</code> flag will convert 'X' to 'N' and
irregular gaps ('.' or '_') to regular gaps '-', and the <code>-r</code> command converts
ambiguous nucleotides to <code>N</code> (don&#8217;t do this if you intend to translate the
sequence, since <code>smof translate</code> will use ambiguous nucleotides to yield an
unambiguous amino acid, when possible).</p>
</li>
<li>
<p><code>smof consensus</code> - Create the consensus sequence of an alignment.</p>
</li>
<li>
<p><code>smof grep -P</code> - Extract entries that match a pattern against the header. For
example, <code>smof grep -P "2020-..-.." myfile.fna</code> will extract entries from the
year 2020.</p>
</li>
<li>
<p><code>smof grep -q</code> - Extract entries or regions that match a pattern in the
sequence. There are many, many options for smof sequence grep and I use most
of them regularly.  For more details, I would recommend you visit the smof
home page.</p>
</li>
</ol>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_motif_extraction">4. Motif extraction</h2>
<div class="sectionbody">
<div class="paragraph">
<p>Amino acid motifs are frequently of interest. Examples include the antigenic
motifs of HA [<a href="#liu2018antigenic">1</a>]. The HA1 and HA2 subunits of the HA gene
can also be thought of as motifs that we may want to extract from a sequences.</p>
</div>
<div class="paragraph">
<p>Motifs may be extracted if we know the motif pattern in a reference gene and
then align this reference against other sequences. Then we can manually trim
the alignment (e.g., in Geneious) to get just the motif for each sequence.</p>
</div>
<div class="paragraph">
<p>This manual approach can be automated, as follows:</p>
</div>
<div class="olist arabic">
<ol class="arabic">
<li>
<p>Align many sequences (along with references).</p>
</li>
<li>
<p>Use a regular expression to find the location of the HA1 region in a subset of the sequences.</p>
</li>
<li>
<p>Extract this location from each sequence in the alignment</p>
</li>
</ol>
</div>
<div class="paragraph">
<p>This approach is very robust against deviation from the expected HA1 pattern
encoded in the regular expression. So long as the expression matches the
correct HA1 region more often than it matches any particular other region, then
the correct region will be extracted.</p>
</div>
<div class="paragraph">
<p>Using MAFFT for alignment and <code>smof</code> for sequence manipulation, this process
can be accomplished as follows:</p>
</div>
<div class="listingblock">
<div class="content">
<pre class="pygments highlight"><code data-lang="bash"><span></span><span class="tok-c1"># translate the longest coding sequence (initial ATG is not required)</span>
smof translate -f flu.fna &gt; flu.faa

<span class="tok-c1"># align the protein sequences</span>
mafft flu.faa &gt; flu.aln

<span class="tok-c1"># extract the positions of $MOTIF as a GFF file</span>
smof grep -qgoP --gff <span class="tok-nv">$MOTIF</span> flu.faa &gt; flu-motif.gff

<span class="tok-c1"># find the most common interval that $MOTIF matches</span>
<span class="tok-nv">bounds</span><span class="tok-o">=</span><span class="tok-k">$(</span>cut -f4,5 flu-motif.gff <span class="tok-p">|</span> sort <span class="tok-p">|</span> uniq -c <span class="tok-p">|</span> sort -rg <span class="tok-p">|</span> head<span class="tok-k">)</span>

<span class="tok-c1"># extract the most common interval from the alignment</span>
smof subseq -b <span class="tok-nv">$bounds</span> flu.aln &gt; motif.faa</code></pre>
</div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_flutile_specialized_tools_for_influenza">5. <code>flutile</code> - specialized tools for influenza</h2>
<div class="sectionbody">
<div class="paragraph">
<p><code>flutile</code> is a python package and command line utility that, like <code>smof</code>, can
be installed through pip: <code>pip install flutile</code> (maybe use <code>pip3</code> on MacOS).</p>
</div>
<div class="sect2">
<h3 id="_ha1_extraction">5.1. HA1 extraction</h3>
<div class="paragraph">
<p>Publicly available HA segment sequences are inconsistent with regards to what
part of the gene the include. Some include DNA before the start codon, some
begin at the start codon, some begin after the start codon. This variety
introduces non-biological noise into phylogenetic studies. Furthermore, the HA
protein is processed after translation to remove the initial signal peptide and
to cut the protein into two subunits, HA1 and HA2. For phylogenetic work, we
usually focus on HA1. For more consistent phylogenetic results, it is good
practice to trim the HA DNA sequences to the part that codes for HA1.</p>
</div>
<div class="paragraph">
<p><code>flutile trim</code> can extract HA1 regions for H1 and H3 sequences (handling for
the other HA subtypes could be easily added given a reference strain and
motifs). <code>flutile</code> finds the HA1 by aligning the input sequences against the
reference sequence (California/07/2009) and then extracting from the alignment
the regions that map to the reference HA1. This will correctly find HA1 regions
even if they are truncated. Here are a few examples:</p>
</div>
<div class="listingblock">
<div class="content">
<pre># DNA input to AA HA1
flutile trim h1-ha1 --conversion=dna2aa myfile.fna
# DNA input to DNA output
flutile trim h1-ha1 --conversion=dna2dna myfile.fna
# AA input to AA output
flutile trim h1-ha1 --conversion=aa2dna myfile.faa</pre>
</div>
</div>
<div class="paragraph">
<p>For tree building, the <code>--conversion=dna2dna</code> option is strongly recommended.</p>
</div>
<div class="paragraph">
<p>To build a gamma tree from the octoFLU reference file, for example, we might use the following command:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>cat reference.fa |  # open the reference file
  smof grep 1A.3.3.3 |  # extract the strains with a gamma HA
  smof grep "|H1|" |  # extract the H1's
  flutile trim h1-ha1 --conversion=dna2dna |  # trim to HA1 region
  mafft /dev/stdin |  # realign the DNA
  FastTree -nt -gtr -gamma &gt; gamma.tre</pre>
</div>
</div>
<div class="paragraph">
<p>The realignment is needed since the alignment <code>flutile</code> makes internally is of
translated inputs against the protein reference.</p>
</div>
</div>
<div class="sect2">
<h3 id="_flutile_and_aadiff_table_generation">5.2. <code>flutile</code> and aadiff table generation</h3>
<div class="paragraph">
<p><code>flutile aadiff</code> takes a multiple-sequence alignment as an input and creates a
character difference table. This command is designed for preparing amino acid
difference tables. Below is an example of a comparison of 4 H1 sequences.</p>
</div>
<div class="paragraph">
<p><code>flutile aadiff --subtype=H1 mufile.faa</code></p>
</div>
<table class="tableblock frame-all grid-all stretch">
<caption class="title">Table 1. <code>flutile aadiff</code> output</caption>
<colgroup>
<col style="width: 20%;">
<col style="width: 20%;">
<col style="width: 20%;">
<col style="width: 20%;">
<col style="width: 20%;">
</colgroup>
<thead>
<tr>
<th class="tableblock halign-left valign-top">site</th>
<th class="tableblock halign-left valign-top">A02479030</th>
<th class="tableblock halign-left valign-top">SD0246</th>
<th class="tableblock halign-left valign-top">SD0272</th>
<th class="tableblock halign-left valign-top">SD0136</th>
</tr>
</thead>
<tbody>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">-3</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A</p></td>
<td class="tableblock halign-left valign-top"></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">T</p></td>
<td class="tableblock halign-left valign-top"></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">-2</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">N</p></td>
<td class="tableblock halign-left valign-top"></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">S</p></td>
<td class="tableblock halign-left valign-top"></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">-1</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">T</p></td>
<td class="tableblock halign-left valign-top"></td>
<td class="tableblock halign-left valign-top"></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">1</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">D</p></td>
<td class="tableblock halign-left valign-top"></td>
<td class="tableblock halign-left valign-top"></td>
<td class="tableblock halign-left valign-top"></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">154</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">K</p></td>
<td class="tableblock halign-left valign-top"></td>
<td class="tableblock halign-left valign-top"></td>
<td class="tableblock halign-left valign-top"></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">154+1</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">-</p></td>
<td class="tableblock halign-left valign-top"></td>
<td class="tableblock halign-left valign-top"></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">X</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">154+2</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">-</p></td>
<td class="tableblock halign-left valign-top"></td>
<td class="tableblock halign-left valign-top"></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">X</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">155</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">D</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">S</p></td>
<td class="tableblock halign-left valign-top"></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">X</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">156</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">D</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">G</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">N</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">X</p></td>
</tr>
</tbody>
</table>
<div class="paragraph">
<p>The <code>--subtype=H1</code> argument tells <code>flutile</code> to align the inputs against an H1
reference (A/United Kingdom/1/1933). The reference is used to determine
relative indices (the <code>sites</code> column). The index reference is used only for
indexing and does not appear in the final table. The first three rows (sites
-3, -2, -1) align to the three residues at the end of the signal peptide. Site
1 is the first residue in the mature peptide. Any gaps in the reference
alignment are indexed as <code>&lt;ref_id&gt;+&lt;offset&gt;</code>, for example 154+1 and 154+2 are
positions 1 and 2 residues after the reference position 154. <code>flutile</code> uses the
references from [<a href="#burke2014">2</a>] (summarized in table below).</p>
</div>
<table class="tableblock frame-all grid-all stretch">
<caption class="title">Table 2. Subtype references, adapted from [<a href="#burke2014">2</a>]</caption>
<colgroup>
<col style="width: 25%;">
<col style="width: 25%;">
<col style="width: 25%;">
<col style="width: 25%;">
</colgroup>
<thead>
<tr>
<th class="tableblock halign-left valign-top">subtype</th>
<th class="tableblock halign-left valign-top">strain name</th>
<th class="tableblock halign-left valign-top">signal peptide</th>
<th class="tableblock halign-left valign-top">mature peptide start</th>
</tr>
</thead>
<tbody>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H1</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A/United Kingdom/1/1933</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">MKARLLVLLCALAATDA</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">DTICIGYHANNS</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H2</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A/Singapore/1/1957</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">MAIIYLILLFTAVRG</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">DQICIGYHANNS</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H3</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A/Aichi/2/1968</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">MKTIIALSYIFCLPLG</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">QDLPGNDNSTATLCLGHHAVPN</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H4</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A/swine/Ontario/019112/1999</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">MLSIAILFLLIAEGSS</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">QNYTGNPVICLGHHAVSN</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H5</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A/Vietnam/1203/2004</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">MEKIVLLFAIVSLVKS</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">DQICIGYHANNS</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H6</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A/chicken/Taiwan/0705/1999</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">MIAIIVIATLAAAGKS</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">DKICIGYHANNS</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H7</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A/Netherlands/219/2003</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">MNTQILVFALVASIPTNA</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">DKICLGHHAVSN</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H8</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A/turkey/Ontario/6118/1968</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">MEKFIAIAMLLASTNA</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">YDRICIGYQSNNS</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H9</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A/swine/Hong Kong/9/1998</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">MEAASLITILLVVTASNA</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">DKICIGYQSTNS</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H10</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A/mallard/bavaria/3/2006</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">MYKIVVIIALLGAVKG</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">LDKICLGHHAVAN</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H11</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A/duck/England/1/1956</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">MEKTLLFAAIFLCVKA</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">DEICIGYLSNNS</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H12</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A/duck/Alberta/60/1976</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">MEKFIILSTVLAASFA</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">YDKICIGYQTNNS</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H13</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A/gull/Maryland/704/1977</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">MALNVIATLTLISVCVHA</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">DRICVGYLSTNS</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H14</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A/mallard/Astrakhan/263/1982</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">MIALILVALALSHTAYS</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">QITNGTTGNPIICLGHHAVEN</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H15</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A/duck/Australia/341/1983</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">MNTQIIVILVLGLSMVRS</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">DKICLGHHAVAN</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H16</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A/black-headed-gull/Turkmenistan/13/1976</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">MMIKVLYFLIIVLGRYSKA</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">DKICIGYLSNNS</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H17</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A/little-yellow-shouldered bat/Guatemala/060/2010</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">MELIILLILLNPYTFVLG</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">DRICIGYQANQN</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H18</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">A/flat-faced bat/Peru/033/2010</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">MITILILVLPIVVG</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">DQICIGYHSNNS</p></td>
</tr>
</tbody>
</table>
<div class="ulist">
<ul>
<li>
<p>The H3 signal peptide appears to actually be <code>MKTIIALSYIFCLALG</code></p>
</li>
</ul>
</div>
<table class="tableblock frame-all grid-all stretch">
<caption class="title">Table 3. Annotation sets</caption>
<colgroup>
<col style="width: 25%;">
<col style="width: 25%;">
<col style="width: 25%;">
<col style="width: 25%;">
</colgroup>
<tbody>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">subtype</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">reference</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">flag</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">description</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H1</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">[<a href="#canton1982">3</a>]</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock"><code>--caton1982</code></p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">antigenic sites</p></td>
</tr>
<tr>
<td class="tableblock halign-left valign-top"><p class="tableblock">H3</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">[<a href="#wiley1981">4</a>]</p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock"><code>--wiley1981</code></p></td>
<td class="tableblock halign-left valign-top"><p class="tableblock">antigenic sites</p></td>
</tr>
</tbody>
</table>
</div>
<div class="sect2">
<h3 id="_flutile_and_indexing">5.3. <code>flutile</code> and indexing</h3>
<div class="paragraph">
<p>The reference strains selected by [<a href="#burke2014">2</a>] may be used for indexing HA,
but no comparable study has been done for NA or the internal genes.</p>
</div>
<div class="paragraph">
<p>Indels are rather rare in the internal genes (with the interesting exception of
NS), so the genes are covered well with a single reference.  I chose the H1N1
A/Puerto_Rico/8/1934 (EPI_ISL_146807) for this role since it 1) has a complete
genome sequence with no ambiguous residues and 2) has been the subject of
extensive structural research.</p>
</div>
<div class="paragraph">
<p>I retrieved all available public WGS data including around 70000 to
80000 sequences for each internal gene. For each segment (other than NS)
the 1st, 2nd and 3rd quantiles were all equal (indicating at least 50%
of the strains in each group are of exactly the same length). The
lengths of the translated A/Puerto_Rico/8/1934 segments exactly matched
these quantiles. For NS, the translated product matched the 2nd quantile
value (the median).</p>
</div>
<div class="paragraph">
<p>So overall, while A/Puerto_Rico/8/1934 may be very different in sequence from
modern strains, it will work fine as an index reference.</p>
</div>
<div class="paragraph">
<p>The NA&#8217;s are a bit more complicated. The NA length varies dramatically even
within subtypes, particularly with variation in length of the stalk-motif
[<a href="#zhou2009special">5</a>, <a href="#castrucci1993biologic">6</a>].</p>
</div>
</div>
<div class="sect2">
<h3 id="_flutile_and_motif_extraction">5.4. <code>flutile</code> and motif extraction</h3>
<div class="paragraph">
<p><code>flutile</code> can also use these subtype reference to extract HA motifs relative to
Burke2014 numbering.</p>
</div>
<div class="paragraph">
<p>For example, to exact the classical H1 antigenic motifs:</p>
</div>
<div class="listingblock">
<div class="content">
<pre>flutile trim motif --conversion=dna2aa --subtype=H1 \
  -m "Sa=124-125,153-157,159-164" \
  -m "Sb=184-195" \
  -m "Ca1=166-170,203-205,235-237" \
  -m "Ca2=137-142,221-222" \
  -m "Cb=70-75" myseqs.fna</pre>
</div>
</div>
</div>
</div>
</div>
<h1 id="_data_visualization" class="sect0">Data Visualization</h1>
<div class="sect1">
<h2 id="_octoflushow">6. octoflushow</h2>
<div class="sectionbody">

</div>
</div>
<h1 id="_clade_inference" class="sect0">Clade inference</h1>
<div class="sect1">
<h2 id="_octoflu">7. octoFLU</h2>
<div class="sectionbody">

</div>
</div>
<h1 id="_standard_reports" class="sect0">Standard Reports</h1>
<div class="sect1">
<h2 id="_quarterly_report">8. quarterly report</h2>
<div class="sectionbody">

</div>
</div>
<h1 id="_whole_genome_sequencing_monthly_selection" class="sect0">Whole Genome Sequencing monthly selection</h1>
<div class="sect1">
<h2 id="_offlu">9. offlu</h2>
<div class="sectionbody">

</div>
</div>
<h1 id="_literature" class="sect0">Literature</h1>
<div class="openblock partintro">
<div class="content">
Add important papers and short summaries
</div>
</div>
<div class="sect1">
<h2 id="_biology_papers">10. Biology papers</h2>
<div class="sectionbody">

</div>
</div>
<div class="sect1">
<h2 id="_method_papers">11. Method papers</h2>
<div class="sectionbody">

</div>
</div>
<div class="sect1">
<h2 id="_glossary">Glossary</h2>
<div class="sectionbody">
<div class="paragraph">
<p>Add any terms likely to be unfamiliar to new team members</p>
</div>
</div>
</div>
<div class="sect1">
<h2 id="_references">References</h2>
<div class="sectionbody">
<div class="paragraph">
<p><a id="liu2018antigenic"></a>[1] S. T. H. Liu, M. A. Behzadi, W. Sun, A. W. Freyn, W.-C. Liu, F. Broecker, R. A. Albrecht, N. M. Bouvier, V. Simon, R. Nachbagauer, &amp; others, Antigenic sites in influenza H1 hemagglutinin display species-specific immunodominance. <em>The Journal of clinical investigation</em>, <strong>128</strong> (2018) 49924996.</p>
</div>
<div class="paragraph">
<p><a id="burke2014"></a>[2] D. F. Burke &amp; D. J. Smith, A recommended numbering scheme for influenza A HA subtypes. <em>PloS one</em>, <strong>9</strong> (2014) e112302.</p>
</div>
<div class="paragraph">
<p><a id="canton1982"></a>[3] A. J. Canton, The antigenic structure of the influenza virus A/PR/8/34 hemagglutinin (H1 subtype). <em>Cell</em>, <strong>31</strong> (1982) 417427.</p>
</div>
<div class="paragraph">
<p><a id="wiley1981"></a>[4] D. C. Wiley, I. A. Wilson, &amp; J. J. Skehel, Structural identification of the antibody-binding sites of Hong Kong influenza haemagglutinin and their involvement in antigenic variation. <em>Nature</em>, <strong>289</strong> (1981) 373378.</p>
</div>
<div class="paragraph">
<p><a id="zhou2009special"></a>[5] H. Zhou, Z. Yu, Y. Hu, J. Tu, W. Zou, Y. Peng, J. Zhu, Y. Li, A. Zhang, Z. Yu, &amp; others, The special neuraminidase stalk-motif responsible for increased virulence and pathogenesis of H5N1 influenza A virus. <em>PloS one</em>, <strong>4</strong> (2009) e6277.</p>
</div>
<div class="paragraph">
<p><a id="castrucci1993biologic"></a>[6] M. R. Castrucci &amp; Y. O. S. H. I. H. I. R. O. Kawaoka, Biologic importance of neuraminidase stalk length in influenza A virus. <em>Journal of virology</em>, <strong>67</strong> (1993) 759764.</p>
</div>
</div>
</div>
</div>
<div id="footer">
<div id="footer-text">
Last updated 2021-04-14 15:20:34 -0500
</div>
</div>
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  messageStyle: "none",
  tex2jax: {
    inlineMath: [["\\(", "\\)"]],
    displayMath: [["\\[", "\\]"]],
    ignoreClass: "nostem|nolatexmath"
  },
  asciimath2jax: {
    delimiters: [["\\$", "\\$"]],
    ignoreClass: "nostem|noasciimath"
  },
  TeX: { equationNumbers: { autoNumber: "none" } }
})
MathJax.Hub.Register.StartupHook("AsciiMath Jax Ready", function () {
  MathJax.InputJax.AsciiMath.postfilterHooks.Add(function (data, node) {
    if ((node = data.script.parentNode) && (node = node.parentNode) && node.classList.contains("stemblock")) {
      data.math.root.display = "block"
    }
    return data
  })
})
</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js?config=TeX-MML-AM_HTMLorMML"></script>
</body>
</html>